### ISSUE TITLE
Indy fails on some parts of Ruby on Rails

### ISSUE THREAD

**Issue by Mahaswami** [jruby/jruby#6191]
**Environment Information**

JRuby 9.2.11 / Ubuntu 18.04 / JDK 14
  
Flags: -Djruby.compile.invokedynamic=true
Rails: 4.2.x

**Expected Behavior**

Rails app to function normally with Indy turned on

**Actual Behavior**

When this flag is turned on, the app blows up with the following exception

Java::ComHeadiusInvokebinder::InvalidTransformException (arguments matching arg.* are not all of the same type):

Please see the attached trace.

[jruby indy rails exception.txt](https://github.com/jruby/jruby/files/4531526/jruby.indy.rails.exception.txt)

We also receive one more exception which we believe Indy related

```
gems/activerecord-jdbc-adapter-1.3.23/lib/arjdbc/jdbc/adapter.rb:528:in `block in exec_query_raw': ArgumentError: wrong number of arguments (given 1, expected 2..3):         SELECT a.attname, format_type(a.atttypid, a.atttypmod),
               pg_get_expr(d.adbin, d.adrelid), a.attnotnull, a.atttypid, a.atttypmod
          FROM pg_attribute a LEFT JOIN pg_attrdef d
            ON a.attrelid = d.adrelid AND a.attnum = d.adnum
         WHERE a.attrelid = '"cg_community"."denormalized_activity_streams"'::regclass
           AND a.attnum > 0 AND NOT a.attisdropped
         ORDER BY a.attnum
 (ActiveRecord::StatementInvalid)
    from gems/activerecord-4.2.8/lib/active_record/connection_adapters/abstract_adapter.rb:484:in `block in log'
    from gems/activesupport-4.2.8/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
```

[jruby indy rails another exception.txt](https://github.com/jruby/jruby/files/4531549/jruby.indy.rails.another.exception.txt)

Though exception points to a SQL, this could be because of some internal Indy wiring is behaving diffeerntly.

**Comment by headius** [comment]
The first issue was fixed after 9.2.11 was released. See #6130 

I don't believe there's any workaround for this. The `method_missing` optimization was applied in JRuby 9.2.11 and I did not make it possible to disable it.

@enebo Perhaps we need a mechanism, either through Options or something finer-grained and more hidden, to be able to tweak all indy optimizations?

@Mahaswami Sorry about this one and the lack of a workaround. If we'd done a 9.2.12 it would be fixed, but we are trying to focus on getting 9.3 done.

**Comment by headius** [comment]
See #6192 for merging the "invalid transform" fix to jruby-9.2 branch.

**Comment by Mahaswami** [comment]
We tried this patch and have more information:

1. TransformError went away.
2. ArgumentError remains.

We have more information about ArgumentError

1. We have one more example we suspect belongs to same root cause. [take_while_exception.txt](https://github.com/jruby/jruby/files/4553097/take_while_exception.txt)
2. We have been able to reproduce with a simpler script as well. 
[invoke_dynamic_exception_reproduce.txt](https://github.com/jruby/jruby/files/4553104/invoke_dynamic_exception_reproduce.txt)

**Comment by headius** [comment]
I believe the ArgumentError may be a bug in ar-jdbc 1.3.x exposed by indy.

Looking at the implementation of `execute_query_raw` in ar-jdbc:

https://github.com/jruby/activerecord-jdbc-adapter/blob/1-3-stable/src/java/arjdbc/jdbc/RubyJdbcConnection.java#L771-L788

We can see here that one definition sets up `required = 1` and the other sets up `required = 2`. When we have bindings that conflict on the number of required arguments like this, we may end up with unpredictable behavior.

In your case, when invokedynamic is enabled, we call along the `IRubyObject[]` path (at least until the indy call optimizes) which requires two arguments. This is similar to other bugs we've seen where the varargs form of a "native" method does not handle the smaller arities.

The code below should patch this issue for ar-jdbc 1.3 but I have not worked on that branch in a long time. I'd need assistance from @kares or @enebo  to test and/or release such a change.

```diff
diff --git a/src/java/arjdbc/jdbc/RubyJdbcConnection.java b/src/java/arjdbc/jdbc/RubyJdbcConnection.java
index ccd789f5..c572fe79 100644
--- a/src/java/arjdbc/jdbc/RubyJdbcConnection.java
+++ b/src/java/arjdbc/jdbc/RubyJdbcConnection.java
@@ -768,7 +768,7 @@ public class RubyJdbcConnection extends RubyObject {
      * @throws SQLException
      * @see #execute_query_raw(ThreadContext, IRubyObject[], Block)
      */
-    @JRubyMethod(name = "execute_query_raw", required = 1) // optional block
+    @JRubyMethod(name = "execute_query_raw") // optional block
     public IRubyObject execute_query_raw(final ThreadContext context,
         final IRubyObject sql, final Block block) throws SQLException {
         final String query = sql.convertToString().getUnicodeValue();
@@ -783,10 +783,12 @@ public class RubyJdbcConnection extends RubyObject {
      * @return raw query result as a name => value Hash (unless block given)
      * @throws SQLException
      */
-    @JRubyMethod(name = "execute_query_raw", required = 2, optional = 1)
-    // @JRubyMethod(name = "execute_query_raw", required = 1, optional = 2)
+    @JRubyMethod(name = "execute_query_raw", required = 1, optional = 2)
     public IRubyObject execute_query_raw(final ThreadContext context,
         final IRubyObject[] args, final Block block) throws SQLException {
+
+        if (args.length == 1) return execute_query_raw(context, args[0], block);
+
         // args: (sql), (sql, max_rows), (sql, binds), (sql, max_rows, binds)
         final String query = args[0].convertToString().getUnicodeValue(); // sql
         IRubyObject max_rows = args.length > 1 ? args[1] : null;
```

**Comment by headius** [comment]
Note that the ArgumentError issue appears to be fixed as of ar-jdbc 50, as part of the refactoring that we did to catch up compatibility. The commit is by @enebo here:

https://github.com/jruby/activerecord-jdbc-adapter/commit/d533cde9525be515d10502e8359d63ee0b4f9b0e

**Comment by headius** [comment]
I have filed jruby/activerecord-jdbc-adapter#1063 for the additional issue, which is not a bug in JRuby (or at least not really). I also filed #6204 to discuss how we can do a better job of catching these bugs before they happen.

I'm going to call this resolved.

**Comment by headius** [comment]
@enebo The only JRuby issue here has been fixed for 9.2 via #6192 and was fixed on master in #6130. I've marked this "Non-Release" the same as #6192.
